### PR TITLE
Ability to catch server errors and defects in Http

### LINKED ISSUE #2056: `decoderResult` is not honored before executing the HttpApp
**Describe the bug**
When the client sends a request with the header size larger than the configured `maxHeaderSize`, netty will get a failure DecoderResult like `failure(io.netty.handler.codec.http.TooLongHttpHeaderException: HTTP header is larger than 1 bytes.)`, 
but the ZIO Http's HttpApp will still be executed and gets an empty request body.

**To Reproduce**
Steps to reproduce the behaviour:

```scala

import zio.Scope
import zio.http._
import zio.http.model._
import zio.test._

object NettyMaxHeaderLengthTest extends ZIOSpecDefault {
  override def spec: Spec[TestEnvironment with Scope, Any] =
    test("should get a failure instead of an empty body") {
      for {
        port <- Server.install(
                  Http
                    .collectZIO[Request] { request =>
                      request.body.asString.map { body =>
                        val responseBody = if (body.isEmpty) "<empty>" else body
                        Response.text(responseBody)
                      } // this should not be run, as the request is invalid
                    }
                    .withDefaultErrorResponse // it should execute the logic within this line
                )
        url     = s"http://localhost:$port"
        headers = Headers.userAgent("a looooooooooooooooooooooooooooog header")

        res  <- Client.request(url, headers = headers, content = Body.fromString("some-body"))
        data <- res.body.asString
      } yield {
        assertTrue(res.status == Status.InternalServerError && data == "<empty>")
      }
    }.provide(
      Client.default,
      Server.live,
      ServerConfig.live(ServerConfig(maxHeaderSize = 1))
    )
}

```

**Expected behaviour**
As the request is invalid, the HttpApp should not be executed. And the error handling logic should be executed (the `withDeafultErrorResponse` in the above snippet)

**Screenshots**
image

**Additional context**
This above snippet was run with zio-http `0.0.4` and `0.0.5`. But it can also be reproduced in the earlier versions(e.g., `2.0.0-RC10`).

### LINKED ISSUE #2065: allow users to handle the decoding errors
**Is your feature request related to a problem? Please describe.**
As discussed in #2063, we would need a better way to allow users to handle the decoding errors, so that they can return different response based on their needs, which is not possible at the moment.

**Describe the solution you'd like**
Having new combinators to handle Http-intrinsic errors, e.g., a decoding error due to the header being too large.

**Describe alternatives you've considered**
(Thanks @vigoo for summarizing the alternatives)
1. Update `Http` to be modelled by `Either[DecodingError, Response] => ZIO[R, Response, Response]`
2.  model decoding errors in the `Request` class and do the check within the Http app
3. be able to produce a `Response` in the exception callback of the server.

### LINKED ISSUE #2071: Add principled, modular means of handling server errors
Conceptually, an `Http` will be modified as:

```scala
Either[Cause[Nothing], In] => ZIO[Env, Err, Out]
```

Practically, however, we will do it by a pair of functions, to avoid boxing:

```scala
serverErrorHandler: Cause[Nothing]=> ZIO[Env, Err, Out]
inputHandler: In => ZIO[Env, Err, Out]
```

When trying to decide how to implement the primitive operators, we look to the conceptual model (the first one, above).

This will allow us to, inside `Http` (rather than `Server`), expose server errors, decoding errors, HTTP composition errors (e.g. NPE in wiring up two Http), and so forth, each `Http` handling such unexpected / fatal / server errors in its own modular way.

### PR DESCRIPTION
**Pull request by vigoo**
Resolves #2071
Resolves #2056
Resolves #2065

### CODE REVIEW

File: zio-http/src/main/scala/zio/http/middleware/Auth.scala [hunk 1]
```diff
@@ -150,19 +150,26 @@ private[zio] trait Auth {
...
-                  Http.fromHttpZIO[Request] { in =>
-                    route
-                      .asInstanceOf[Http.Route[R1, Err1, Request, Response]]
-                      .run(in)
-                      .provideSomeEnvironment[R0](_.union[Context](ZEnvironment(context)))
-                      .map { (http: Http[R1, Err1, Request, Response]) =>
-                        self.apply(http)
-                      }
-                  }
+                case empty @ Http.Empty(_)              => empty.asInstanceOf[Http.Empty[R0, Response]]
```
Comment (vigoo) [inline zio-http/src/main/scala/zio/http/middleware/Auth.scala]: These casts in this file are probably invalid and we need to provide environment for the error handler too - to be checked

File: zio-http/src/main/scala/zio/http/Http.scala [hunk 2]
```diff
@@ -36,14 +36,20 @@ sealed trait Http[-R, +Err, -In, +Out] { self =>
    */
   final def >>>[R1 <: R, Err1 >: Err, In1 >: Out, Out1](
     handler: Handler[R1, Err1, In1, Out1],
+    errorOutputMapper: Out => Out1,
```
Comment (vigoo) [inline zio-http/src/main/scala/zio/http/Http.scala:39]: This operator became weird by this parameter, and in fact - at least in zio-http's codebase - every time it is used, `Out1` is `Unit`. So maybe we should replace it with a stricter, single parameter version.

File: zio-http/src/main/scala/zio/http/Http.scala [hunk 3]
```diff
@@ -70,33 +76,46 @@ sealed trait Http[-R, +Err, -In, +Out] { self =>
...
         new Route[R1, Err1, In1, Out1] {
           override def run(in: In1): ZIO[R1, Err1, Http[R1, Err1, In1, Out1]] =
             route.run(in).map(_.defaultWith(that))
+
+          override val errorHandler: Option[Cause[Nothing] => ZIO[R1, Nothing, Out1]] =
+            route.errorHandler
         }
     }
 
+  private[http] def errorHandler: Option[Cause[Nothing] => ZIO[R, Nothing, Out]]
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/Http.scala:142]: Why put it into an Option?

Comment (vigoo) [reply zio-http/src/main/scala/zio/http/Http.scala:142]: We cannot have a default error handler producing arbitrary Out. So having one is optional and on the edge where it is always known to be Response we can choose a default handler if needed

Comment (vigoo) [reply zio-http/src/main/scala/zio/http/Http.scala:142]: And also this way we don’t have to catchAll ZIOs on each level just where we specify the handler

File: zio-http/src/main/scala/zio/http/Http.scala (hunk 2 above)
Comment (jdegoes) [reply zio-http/src/main/scala/zio/http/Http.scala:39]: Yes, I agree. This is a bit too weird!

File: zio-http/src/main/scala/zio/http/Http.scala [hunk 4]
```diff
@@ -70,33 +96,46 @@ sealed trait Http[-R, +Err, -In, +Out] { self =>
   )(implicit trace: Trace): Http[R1, Err1, In1, Out1] =
     self.defaultWith(that)
 
+  final def catchAllDefect[R1 <: R, Out1 >: Out](f: Cause[Nothing] => ZIO[R1, Nothing, Out1])(implicit
```
Comment (jdegoes) [inline zio-http/src/main/scala/zio/http/Http.scala]: I think `catchAllCause` should let you get the defects. No need for a `catchAllDefect` method in particular.

Comment (vigoo) [reply zio-http/src/main/scala/zio/http/Http.scala]: Fixed

**Review by jdegoes — APPROVED** [review]

**Pull request merged**